float32 w_left
float32 w_right